src/parser.coffee

```diff
@@ -225,47 +226,56 @@ class exports.Parser extends events.EventEmitter
       if s
         s.cdata = true
 
-  parseString: (str, cb) =>
-    if cb? and typeof cb is "function"
-      @on "end", (result) ->
-        @reset()
-        cb null, result
-      @on "error", (err) ->
-        @reset()
-        cb err
+  setupCallback: (event, callbacks) =>
+    @on event, (result) =>
+      @reset()
+      errToThrow = null
+      for cb in callbacks
+        try
+            cb result
+        catch err
+          errToThrow = err
+      if errToThrow
+        throw errToThrow
 
-    try
-      str = str.toString()
-      if str.trim() is ''
-        @emit "end", null
-        return true
-
-      str = bom.stripBOM str
-      if @options.async
-        @remaining = str
-        setImmediate @processAsync
-        return @saxParser
-      @saxParser.write(str).close()
-    catch err
-      unless @saxParser.errThrown or @saxParser.ended
-        @emit 'error', err
-        @saxParser.errThrown = true
-      else if @saxParser.ended
-        throw err
+  parseString: (str, cb) =>
+    new Promise (resolve, reject) =>
+      if not (cb? and typeof cb is "function")
+        cb = () ->
+      @setupCallback "end", [resolve, (r) -> (cb null, r)]
+      @setupCallback "error", [reject, cb]
+
+      try
+        str = str.toString()
+        if str.trim() is ''
+          @emit "end", null
+          return true
+
+        str = bom.stripBOM str
+        if @options.async
+          @remaining = str
+          setImmediate @processAsync
+          return @saxParser
+        @saxParser.write(str).close()
+      catch err
+        unless @saxParser.errThrown or @saxParser.ended
+          @emit 'error', err
+          @saxParser.errThrown = true
+        else if @saxParser.ended
+          throw err
+    .catch(()->) # disable unhandled promise rejection warnings
 
 exports.parseString = (str, a, b) ->
   # let's determine what we got as arguments
-  if b?
-    if typeof b == 'function'
-      cb = b
-    if typeof a == 'object'
-      options = a
-  else
-    # well, b is not set, so a has to be a callback
-    if typeof a == 'function'
-      cb = a
-    # and options should be empty - default
-    options = {}
+  options = {}
+  if a? and typeof a == 'function'
+    cb = a
+  if a? and typeof a == 'object'
+    options = a
+  if b? and typeof b == 'function'
+    cb = b
+  if b? and typeof b == 'object'
+    options = b
```

Comment:
I am a bit confused why `b` should accept an object and set this to the options. That way you can set `a` to the options and `b` to another set of options and `b` "wins".

Reply:
Yes, you are right